### PR TITLE
Enhance: switch to rollup

### LINKED ISSUE #10: Switch to a better toolkit
The problem with browserify is that it uses insert-module-globals which always wraps modules into an additional IIFE + call. It also doesn't seem to have a straightforward method of applying our transforms on-the-fly to child packages, meaning that we have to patch stylelint on the disk, and that is wasteful, slow, and just wrong.

I guess we'll try rollup.

* Add stylelint as a dependency in package.json.
* Apply transforms on-the-fly.
* Rewrite some patches to use dummy modules where possible and reasonable, but keep regexp if there's a noticeable size reduction, say at least half a kb.
* Keep the check to ensure each replacement actually occurred, and call process.exit(1) otherwise.

### PR DESCRIPTION
**Pull request by eight04**
Fixes #10 

* [x] Currently this is blocked by a plugin bug: https://github.com/rollup/plugins/issues/1235, or we can just add a replace rule like `/process\.cwd\(\)/` -> `"/"`.
* [x] This branch doesn't include surgarss. Maybe we should switch to less/stylus-lang parser instead?

### CODE REVIEW

**Comment by tophf** [comment]
Re "less/stylus-lang parser", I've tried building post-styl but it includes stylus-lang so it's quite big and is a waste as we already have one. Ideally Stylus could expose it via its own `require` I guess, but I'd say including sugarss is reasonable because it's so small.

File: rollup.config.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    alias({
+      entries: {
+        util: require.resolve("./shim/util"),
+        tty: require.resolve("./shim/tty"),
+        os: require.resolve("./shim/os"),
+        [require.resolve("stylelint/lib/getConfigForFile")]: require.resolve("./shim/getConfigForFile"),
+        [require.resolve("stylelint/lib/isPathIgnored")]: require.resolve("./shim/isPathIgnored"),
+        ...Object.fromEntries([
+          "cosmiconfig",
+          "css-functions-list",
```
Comment (tophf) [inline rollup.config.js:59]: I thought this list helps some meaningful rule to warn about unknown css functions.

**Review by tophf — APPROVED** [review]
Awesome work, you shaved off 20% and another 100kB! Don't forget to add yourself to `authors` in package.json :-)

**Comment by tophf** [comment]
Er, one thing is super wrong though: it doesn't fail the build when a pattern wasn't found.

File: rollup.config.js (hunk 1 above)
Comment (eight04) [reply rollup.config.js:59]: I thought it was included via `rollup-plugin-re` but seems that it was not.

**Comment by eight04** [comment]
https://github.com/jetiny/rollup-plugin-re/issues/14

**Comment by tophf** [comment]
Hmm, sorry, it looks like css-functions-unknown was already handled as an inline injection, so maybe there was no need to include it as a dependency?

**Comment by eight04** [comment]
Yeah. I just fixed the regex pattern, by replacing forward slash with dot, to match Windows' path separator (`\`).

**Comment by tophf** [comment]
I've tried polyfilling `testOrThrow` but it didn't stop the build, so I've reverted it. 

```js
/** @this {RegExp} */
function testOrThrow(str) {
  if (!RegExp.prototype.test.call(this, str)) {
    console.error('Not found: ' + this.source);
    process.exit(1);
  }
  return true;
}

function mustMatch(patterns) {
  for (const p of patterns) {
    if (typeof p.test === 'string') {
      p.test = new RegExp(p.test.replace(/[{}()[\]\\.+*?^$|]/g, '\\$&'));
    }
    if (p.test instanceof RegExp) {
      p.test.test = testOrThrow;
    }
  }
  return patterns;
}
```
```js
    re({
      patterns: mustMatch([
```

**Comment by eight04** [comment]
Done.

File: README.md [hunk 2]
```diff
@@ -40,16 +31,15 @@ The resulting bundle:
...
-* Use CommonJS `require` or AMD `define` or a global `stylelint` (if you don't use modules) to load the bundled script, then access the `lint` function:
+* Access the `stylelint` global:
 
   ```js
-  // const stylelint = require("stylelint"); // uncomment if you use modules
   stylelint.lint({
     code: "body { color: #000; }",
     config: {
-      syntax: 'sugarss',
+      customSyntax: 'sugarss',
```
Comment (eight04) [inline README.md:40]: I'm not really sure if this was a documentation error or we did use a wrong property to define custom syntax.

Comment (tophf) [reply README.md:40]: It's an intentionally different name I've used to set the syntax, IIRC. I should have probably used `__openstyles__syntax` or something like that

Comment (tophf) [reply README.md:40]: https://github.com/openstyles/stylus/blob/030621462cde7fcc773eb3f627e64fb57a564422/edit/editor-worker.js#L73-L75

**Comment by tophf** [comment]
I've tried it in Stylus and there was an error about stylelint.SugarSSParser.prototype.checkSemicolon - previously I exposed sugarss as `SugarSSParser` in order to patch the error in our editor-worker. I've commented this out, but it didn't help anyway, the linter didn't find anything wrong with obviously broken CSS.

**Comment by eight04** [comment]
Yeah I think sugarss doesn't allow semicolon. Should we

1. Just remove sugarss?
2. Expose sugarss parser to the `stylelint` object so it can be patched?

**Comment by tophf** [comment]
Stylus currently does 2.

**Comment by eight04** [comment]
Done.

**Comment by tophf** [comment]
You didn't expose the parser class, so edit-worker in Stylus still fails.

**Comment by eight04** [comment]
I thought the entire module were included. How do you access the parser class?

**Comment by tophf** [comment]
https://github.com/openstyles/stylus/blob/master/edit/editor-worker.js#L69

**Comment by eight04** [comment]
Done

**Comment by tophf** [comment]
Yep, but it didn't help with the bigger second issue: the linter still doesn't report any problems with obviously broken CSS.

**Comment by tophf** [comment]
The demo works, but nothing happens in Stylus extension when I replace stylelint-bundle.min.js with the new one from here and change `color: red` to `color2: red`.

**Comment by tophf** [comment]
Did you even test the result in Stylus?

**Comment by tophf** [comment]
I've debugged it myself: `stylelint.createLinter(...)._lintSource is not a function`
I'll investigate...

**Comment by tophf** [comment]
So, previously I was using the internal API because I couldn't find a reliable method of determining the necessity of enabling sugarss mode. Looks like there's no need for that anymore, at least when editing an existing style.

Assuming it now fails properly when a pattern wasn't found in the source code, everything's fine.

**Pull request merged**